source/simulator/introspection.cc

```diff
@@ -433,31 +436,19 @@ namespace aspect
 
 
   template <int dim>
-  const std::vector<unsigned int>
+  const std::vector<unsigned int> &
   Introspection<dim>::get_indices_for_fields_of_type (const CompositionalFieldDescription::Type &type) const
   {
-    std::vector<unsigned int> indices;
-
-    for (unsigned int i=0; i<n_compositional_fields; ++i)
-      if (composition_descriptions[i].type == type)
-        indices.push_back(i);
-
-    return indices;
+    return composition_indices_for_type[type];
```

Comment:
Here and in the other places, can you add an assertion that `type` is a valid index?